### PR TITLE
IC-1352 Created Appointments Database schema and repo

### PR DESCRIPTION
**Pull request by mahalrmoj**
## What does this pull request do?

Create the database structures necessary to implement sessions and appointments first cut functionality.

## What is the intent behind these changes?

Provide the persistence tier on which to build the API functionality

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  // Activities
+  var appointmentTime: OffsetDateTime? = null,
+  var durationInMinutes: Int? = null,
+
+  // Status
+  @NotNull @ManyToOne @Fetch(FetchMode.JOIN) val createdBy: AuthUser,
+  @NotNull val createdAt: OffsetDateTime,
+
+  // Required
+  @NotNull @ManyToOne val actionPlan: ActionPlan,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt:29]: do we want to specify the fetchtype here? it's likely we will not want the full action plan when looking up appointments? so lazy?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlan.kt [hunk 2]
```diff
@@ -32,6 +33,10 @@ data class ActionPlan(
   @ManyToOne @Fetch(FetchMode.JOIN) var submittedBy: AuthUser? = null,
   var submittedAt: OffsetDateTime? = null,
 
+  // This isn't set directly and hence it's a 'val'. Once an action plan appointment has been persisted, on retrieval
+  // of the associated action plan this list will be automatically set by hibernate.
+  @OneToMany(mappedBy = "actionPlan") val appointments: List<ActionPlanAppointment>,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlan.kt:38]: i read somewhere that using `Set` results in more efficient SQL than `List` https://vladmihalcea.com/hibernate-facts-favoring-sets-vs-bags/

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt (hunk 1 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt:29]: In "maintaining" appointments, I'm not sure we need to eagerly load action plans (and joined activities) as we don't use anything from action plan - so the action plan wont need to get hydrated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlan.kt (hunk 2 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlan.kt]: On reflection a bi-directional relationship is not necessary. At present, an action plan doesn't need to know about its appointments as it's not part of the action plan dto returned to the ui. Appointments can be retrieved directly given the requirements of the current set of endpoints. So I have removed the list (which should have been a set) for now.

**Review by tomsmyers — APPROVED** [review]

**Pull request merged**